espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/Element.java

```diff
@@ -68,46 +85,67 @@ public static ViewInteraction getViewInteractionById(String elementId) throws Ap
-                }
+    public static boolean exists(String elementId) {
+        return cache.containsKey(elementId);
+    }
 
-                return view;
-            }
-        };
+    public static View getCachedView(String elementId) throws NoSuchElementException, StaleElementException {
+        if (!exists(elementId)) {
+            throw new NoSuchElementException(String.format("No such element with ID %s", elementId));
+        }
 
-        return new UiControllerPerformer<>(runnable).run();
-    }
+        View view = cache.get(elementId);
 
-    public static View getViewById(String elementId, boolean skipWaitForMainThreadIdle) throws AppiumException {
-        if (skipWaitForMainThreadIdle) {
-            if (!exists(elementId)) {
-                throw new NoSuchElementException(String.format("Invalid element ID %s", elementId));
-            }
-            View view = cache.get(elementId);
+        if (!view.isShown()) {
+            throw new StaleElementException(elementId);
+        }
+
+        // This is a special case:
+        //
+        // If the contentDescription of a cached view changed, it almost certainly means the rendering of an
+        // AdapterView (ListView, GridView, Scroll, etc...) changed and the View contents were shuffled around.
+        //
+        // If we encounter this, try to scroll the element with the expected contentDescription into
+        // view and return that element
+        if (contentDescriptionCache.containsKey(elementId)) {
+            String expectedContentDesc = contentDescriptionCache.get(elementId);
+
+            if (view.getContentDescription() != null && !view.getContentDescription().equals(expectedContentDesc)) {
+
+                // Look up the view hierarchy to find the closest ancestor AdapterView
+                ViewParent ancestorAdapter = view.getParent();
+                while (ancestorAdapter != null && !(ancestorAdapter instanceof AdapterView)) {
+                    ancestorAdapter = ancestorAdapter.getParent();
+                }
 
-            if (!view.isShown()) {
-                throw new StaleElementException(elementId);
+                try {
+                    // Try scrolling the view with the expected content description into the viewport
+                    DataInteraction dataInteraction = onData(
+                            hasEntry(Matchers.equalTo("contentDescription"), is(expectedContentDesc))
+                    );
+                    if (ancestorAdapter != null) {
+                        dataInteraction.inAdapterView(withView((AdapterView) ancestorAdapter));
+                    }
+                    dataInteraction.check(matches(isDisplayed()));
+
+                    // If successful, use that view instead of the cached view
+                    view = (new ViewGetter()).getView(onView(withContentDescription(expectedContentDesc)));
+                    cache.put(elementId, view);
+                } catch (Exception e) {
```

Comment:
shall we log the original exception?

Reply:
Perhaps it only makes sense to only filter Espresso exceptions instead of the generic one?

Reply:
I tried looking into narrowing down the Exceptions, but the exceptions are all runtime and I don't know which ones to catch.

Reply:
I'd say https://developer.android.com/reference/android/support/test/espresso/EspressoException